File: iteration2/src/Transcript.java
import org.json.simple.JSONArray;
import org.json.simple.JSONObject;
import org.json.simple.parser.JSONParser;

import java.io.FileReader;
import java.util.ArrayList;

public class Transcript {

	private Student student;
	private ArrayList<Double> gano;
	private ArrayList<Semester> semesters;
	Colors colors = new Colors();

	public Transcript(Student student) {
		this.student = student;
		this.gano = new ArrayList<Double>();
		this.semesters = new ArrayList<Semester>();

		createTranscript();
	}

	private void createTranscript() {
		initializeSemesters();
		initializeGanoList();
	}

	private void initializeSemesters() {
		String filePath = "./jsons/student/" + student.getID() + ".json";
		JSONParser parser = new JSONParser();

		try {
			// read student json file
			Object obj = parser.parse(new FileReader(filePath));
			JSONObject studentJson = (JSONObject) obj;
			// get transcript attribute then obtain the array of courses inside
			JSONObject transcript = (JSONObject) studentJson.get("Transcript");
			JSONArray arrayOfSemesters = (JSONArray) transcript.get("Semester");

			// fetch each semester's courses
			for (int i = 0; i < arrayOfSemesters.size(); i++) {
				// get current semester with its corresponding courses
				JSONObject semesterJson = (JSONObject) arrayOfSemesters.get(i);
				JSONArray coursesArray = (JSONArray) semesterJson.get("Courses");
				ArrayList<Course> semesterCourses = new ArrayList<>();

				for (int j = 0; j < coursesArray.size(); j++) {
					JSONObject jsonCourse = (JSONObject) coursesArray.get(j);

					// variable parameters to create course objects
					// get method in json-simple returns object as an Object needs a cast
					String courseName = (String) jsonCourse.get("CourseName");
					String courseID = (String) jsonCourse.get("CourseID");
					int credit = ((Number) jsonCourse.get("Credit")).intValue();
					String type = (String) jsonCourse.get("CourseType");
					int semesterVal = ((Number) jsonCourse.get("Semester")).intValue();
					double grade = ((Number) jsonCourse.get("Grade")).doubleValue();

					Course course = new Course(courseName, courseID, credit, type, semesterVal, grade);
					semesterCourses.add(course);
				}

				// construct Semester objects and initialize other attributes
				Semester semester = new Semester(semesterCourses);

				// Check if "SemesterInf" is present
				if (semesterJson.containsKey("SemesterInf")) {
					// retrieve "SemesterInf" object
					JSONObject semesterInf = (JSONObject) semesterJson.get("SemesterInf");

					semester.setTakenCredit(((Number) semesterInf.get("TakenCredit")).intValue());
					semester.setCompletedCredit(((Number) semesterInf.get("CompletedCredit")).intValue());
					semester.setYano(((Number) semesterInf.get("Yano")).doubleValue());
				}

				semesters.add(semester);
			}

		} catch (Exception exception) {
			exception.printStackTrace();
		}
	}

	// initialize GANO from studentID.json
	private void initializeGanoList() {
		String filePath = "./jsons/student/" + student.getID() + ".json";
		JSONParser parser = new JSONParser();

		try {
			// read student json file
			Object obj = parser.parse(new FileReader(filePath));
			JSONObject studentJson = (JSONObject) obj;
			// get transcript attribute then obtain the array of courses inside
			JSONObject transcript = (JSONObject) studentJson.get("Transcript");
			JSONArray arrayOfSemesters = (JSONArray) transcript.get("Semester");

			// fetch each semester's yano
			for (int i = 0; i < arrayOfSemesters.size(); i++) {
				JSONObject semester = (JSONObject) arrayOfSemesters.get(i);
				// check if null
				if (semester.containsKey("SemesterInf")) {
					JSONObject semesterInf = (JSONObject) semester.get("SemesterInf");

					if (semesterInf.containsKey("Gano")) {
						// since getting a value from JSON returns a long value, cast to number, then
						// get double value
						Double ganoVal = ((Number) semesterInf.get("Gano")).doubleValue();
						gano.add(ganoVal);
					}
				}
			}

		} catch (Exception exception) {
			exception.printStackTrace();
		}
	}

	public void viewTranscript() {
		System.out.println(
				colors.getRED() + colors.getBOLD() + "\n>> Transcript\n" + colors.getRESET() + colors.getRESET());

		System.out.print(colors.getGREEN() + "Student ID: " + colors.getRESET() + student.getID() + "\n"
				+ colors.getGREEN() + "Name and Surname: " + colors.getRESET() + student.getName() + " "
				+ student.getSurname() + "\n");
		for (int i = 0; i < semesters.size(); i++) {
			System.out.println(
					"-------------------------------------------------------------------------------------------------------------");
			System.out.println(colors.getBLUE()+"Semester " + (i + 1) + "\n"+colors.getRESET());
			System.out.print(colors.getYELLOW());
			System.out.printf("\t%-15s%-70s%-10s%-10s%n%n", "Course Code", "Course Name", "Credit", "Grade");
			System.out.print(colors.getRESET());
			Semester currentSemester = semesters.get(i);
			for (Course course : currentSemester.getCourses()) {
				System.out.printf("\t%-15s%-70s%-10s%-10s%n", course.getCourseID(),
						course.getCourseName(), course.getCredit(),
						course.getGrade());
			}
			System.out.println();
			int takenCredit = currentSemester.getTakenCredit();
			if (takenCredit != 0) {
				System.out.println(colors.getYELLOW()+"Taken Credit: "+colors.getRESET() + takenCredit);
			}
			System.out.println(colors.getYELLOW()+"Completed Credit: "+colors.getRESET()
					+ (currentSemester.getCompletedCredit() != 0 ? currentSemester.getCompletedCredit()
							: "Semester has not been completed yet"));
			System.out.println(colors.getYELLOW()+"Yano: "+colors.getRESET()
					+ (currentSemester.getYano() != 0.0 ? String.format("%.2f", currentSemester.getYano())
							: "Semester has not been completed yet"));

			int ganoIndex = i < gano.size() ? i : gano.size() - 1;
			double currentGano = ganoIndex >= 0 ? gano.get(ganoIndex) : 0.0;
			if (currentGano != 0.0) {
				System.out.println(colors.getYELLOW()+"Gano: "+colors.getRESET() + String.format("%.2f", currentGano));
			} else {
				System.out.println("Gano: N/A");
			}
			System.out.println();
		}
	}

	public void setGano(ArrayList<Double> gano) {
		this.gano = gano;
	}

	public ArrayList<Double> getGano() {
		return gano;
	}

	public ArrayList<Semester> getSemesters() {
		return semesters;
	}

	public void setSemesters(ArrayList<Semester> semesters) {
		this.semesters = semesters;
	}
}

File: iteration3/src/transcript.py
import json
from typing import List

from colors import Colors
from semester import Semester
from course import Course


class Transcript:
    def __init__(self, student):
        # private attributes
        self.__student = student
        self.__gano: List[float] = []
        self.__semesters: List[Semester] = []
        self.__colors = Colors()

        # initialization method
        self.__create_transcript()

    def __create_transcript(self):
        # helper methods to initialize semesters and GANO list
        self.__initialize_semesters()
        self.__initialize_gano_list()

    def __initialize_semesters(self):
        # read and initialize semester information from JSON file
        file_path = f"./jsons/student/{self.__student.get_id()}.json"

        try:
            with open(file_path, "r") as file:
                student_json = json.load(file)
                transcript = student_json.get("Transcript", {})
                array_of_semesters = transcript.get("Semester", [])

                for semester_json in array_of_semesters:
                    courses_array = semester_json.get("Courses", [])
                    semester_courses = []

                    for json_course in courses_array:
                        # extracting course information from JSON
                        course_name = json_course.get("CourseName", "")
                        course_id = json_course.get("CourseID", "")
                        credit = int(json_course.get("Credit", 0))
                        course_type = json_course.get("CourseType", "")
                        semester_val = int(json_course.get("Semester", 0))
                        grade = float(json_course.get("Grade", 0.0))

                        # constructing Course objects
                        course = Course(
                            course_name,
                            course_id,
                            credit,
                            course_type,
                            semester_val,
                            grade,
                        )
                        semester_courses.append(course)

                    # creating Semester objects
                    semester = Semester(semester_courses)

                    if "SemesterInf" in semester_json:
                        semester_inf = semester_json["SemesterInf"]
                        semester.set_taken_credit(
                            int(semester_inf.get("TakenCredit", 0))
                        )
                        semester.set_completed_credit(
                            int(semester_inf.get("CompletedCredit", 0))
                        )
                        semester.set_yano(float(semester_inf.get("Yano", 0.0)))

                    self.__semesters.append(semester)

        except Exception as e:
            print(f"Error: {e}")

    def __initialize_gano_list(self):
        # helper method to initialize GANO list from JSON file
        file_path = f"./jsons/student/{self.__student.get_id()}.json"

        try:
            with open(file_path, "r") as file:
                student_json = json.load(file)
                transcript = student_json.get("Transcript", {})
                array_of_semesters = transcript.get("Semester", [])

                for semester in array_of_semesters:
                    if "SemesterInf" in semester:
                        semester_inf = semester["SemesterInf"]

                        if "Gano" in semester_inf:
                            # Extracting GANO values
                            gano_val = float(semester_inf["Gano"])
                            self.__gano.append(gano_val)

        except Exception as e:
            print(f"Error: {e}")

    def view_transcript(self):
        # method to display the transcript information
        print(
            self.__colors.get_red()
            + self.__colors.get_bold()
            + "\n>> Transcript\n"
            + self.__colors.get_reset()
            + self.__colors.get_reset()
        )

        print(
            f"{self.__colors.get_green()}Student ID: {self.__student.get_id()}\n"
            f"Name and Surname: {self.__student.get_name()} {self.__student.get_surname()}{self.__colors.get_reset()}"
        )

        for i, semester in enumerate(self.__semesters):
            # displaying information for each semester
            print(
                "--------------------------------------------------------------------------------------------------------------"
            )
            print(
                f"{self.__colors.get_blue()}Semester {i + 1}{self.__colors.get_reset()}"
            )
            print(self.__colors.get_yellow())
            print("{:<10}{:<15}{:<70}{:<10}{:<10}".format("","Course Code", "Course Name", "Credit", "Grade"))

            print(self.__colors.get_reset())

            for course in semester.get_courses():
                # displaying course information
                print("{:<10}{:<15}{:<70}{:<10}{:<10}".format(
                    "",
                    course.get_course_id(),
                    course.get_course_name(),
                    course.get_credit(),
                    course.get_grade(),
                ))
            
            print()


            taken_credit = semester.get_taken_credit()
            if taken_credit != 0:
                print(
                    f"{self.__colors.get_yellow()}Taken Credit: {self.__colors.get_reset()}{taken_credit}"
                )

            print(
                f"{self.__colors.get_yellow()}Completed Credit: {self.__colors.get_reset()}"
                f"{semester.get_completed_credit() if semester.get_completed_credit() != 0 else 'Semester has not been completed yet'}"
            )

            print(
                f"{self.__colors.get_yellow()}Yano: {self.__colors.get_reset()}"
                f"{f'{semester.get_yano():.2f}' if semester.get_yano() != 0.0 else 'Semester has not been completed yet'}"
            )

            gano_index = min(i, len(self.__gano) - 1)
            current_gano = self.__gano[gano_index] if gano_index >= 0 else 0.0

            if current_gano != 0.0:
                print(
                    f"{self.__colors.get_yellow()}Gano: {self.__colors.get_reset()}{current_gano:.2f}"
                )
            else:
                print("Gano: N/A")

           

    def set_gano(self, gano: List[float]):
        self.__gano = gano

    def get_gano(self) -> List[float]:
        return self.__gano

    def get_semesters(self) -> List[Semester]:
        return self.__semesters

    def set_semesters(self, semesters: List[Semester]):
        self.__semesters = semesters
